File: patch/patch.go
// Package patch loads YAML synth programs and applies them to fresh
// voices. The on-disk format mirrors the Java synth: a top-level map of
// component name -> spec, where each spec has a `type` (which routes to
// a registered component.Factory) and an optional `out` directive
// describing where this component's output is wired.
//
// `out: main` plugs into the voice's main summing junction.
// `out: <name>` plugs into another component's default ("main") input.
// `out: <name>.<select>` plugs into a named input on another component.
package patch

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"

	"ondes/component"
	"ondes/synth"
)

// Patch is a parsed YAML program. It is immutable after Load and can be
// applied to many voices.
type Patch struct {
	name  string
	specs map[string]component.Spec
	// Stable instantiation order so that `out:` resolution is
	// deterministic and Apply produces the same wire topology every time.
	order []string
}

func (p *Patch) Name() string { return p.name }

// Load resolves a patch by name, searching the on-disk `./program/`
// tree first (recursively) and then the bundled patches under
// `src/main/resources/program/`. Substring matching is supported to
// mirror the Java loader: callers can pass `program/sine` to
// disambiguate against `bell-organ` etc.
func Load(name string) (*Patch, error) {
	path, err := resolvePath(name)
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return parse(name, data)
}

func resolvePath(name string) (string, error) {
	candidates, err := collectPaths()
	if err != nil {
		return "", err
	}
	// Exact basename match wins.
	target := name + ".yaml"
	for _, c := range candidates {
		if filepath.Base(c) == target {
			return c, nil
		}
	}
	// Otherwise first substring match (case-insensitive against the
	// path fragment after the program root).
	lname := strings.ToLower(name)
	for _, c := range candidates {
		if strings.Contains(strings.ToLower(c), lname) && strings.HasSuffix(c, ".yaml") {
			return c, nil
		}
	}
	return "", fmt.Errorf("patch %q not found", name)
}

func collectPaths() ([]string, error) {
	var out []string
	for _, root := range []string{"program", filepath.Join("src", "main", "resources", "program")} {
		if _, err := os.Stat(root); err != nil {
			continue
		}
		err := filepath.Walk(root, func(p string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() && strings.HasSuffix(p, ".yaml") {
				out = append(out, p)
			}
			return nil
		})
		if err != nil {
			return nil, err
		}
	}
	sort.Strings(out)
	return out, nil
}

func parse(name string, data []byte) (*Patch, error) {
	var raw map[string]map[string]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("patch %s: %w", name, err)
	}
	p := &Patch{name: name, specs: map[string]component.Spec{}}
	for k, v := range raw {
		p.specs[k] = component.Spec(v)
		p.order = append(p.order, k)
	}
	sort.Strings(p.order)
	return p, nil
}

// Apply instantiates this patch into a fresh voice. It runs in two
// passes: first Make + Configure every component, then resolve `out:`
// directives now that every component (and its Output() wire) exists.
func (p *Patch) Apply(v *synth.Voice) error {
	insts := make(map[string]component.Component, len(p.order))
	for _, name := range p.order {
		spec := p.specs[name]
		typeKey, _ := spec["type"].(string)
		if typeKey == "" {
			return fmt.Errorf("patch %s: component %q missing type", p.name, name)
		}
		c, err := component.Make(typeKey)
		if err != nil {
			return fmt.Errorf("patch %s: component %q: %w", p.name, name, err)
		}
		if err := c.Configure(spec, v, name); err != nil {
			return fmt.Errorf("patch %s: component %q: %w", p.name, name, err)
		}
		insts[name] = c
		v.AddComponent(name, c)
	}
	for _, name := range p.order {
		c := insts[name]
		spec := p.specs[name]
		if raw, ok := spec["out"]; ok {
			dest, ok := raw.(string)
			if !ok {
				return fmt.Errorf("patch %s: component %q: out must be a string", p.name, name)
			}
			if err := wireOut(p.name, name, dest, c.Output(), insts, v); err != nil {
				return err
			}
		}
		// Sidechain outputs: nested-map (linear-out / log-out on
		// midi-note) and flat-string (out-level on env). Wire each
		// according to the component's exposed named output.
		for _, key := range []string{"linear-out", "log-out"} {
			dest, ok := nestedOut(spec, key)
			if !ok {
				continue
			}
			if w := namedOutput(c, key); w != nil {
				if err := wireOut(p.name, name, dest, w, insts, v); err != nil {
					return err
				}
			}
		}
		if dest, ok := spec["out-level"].(string); ok {
			if w := namedOutput(c, "out-level"); w != nil {
				if err := wireOut(p.name, name, dest, w, insts, v); err != nil {
					return err
				}
			}
		}
	}
	return nil
}

// nestedOut is the `linear-out: { amp:..., out: "lpf.freq" }` pattern:
// returns the target named in the block's `out:` field.
func nestedOut(spec component.Spec, key string) (string, bool) {
	m, ok := spec[key].(map[string]any)
	if !ok {
		return "", false
	}
	dest, ok := m["out"].(string)
	return dest, ok
}

// namedOutput returns a sidechain wire from a component, if it exposes one
// under the given pin name. Components opt in by implementing methods of
// the same name as the spec key (LinearOut / LogOut / LevelOut).
func namedOutput(c component.Component, key string) *synth.Wire {
	type linearOuter interface{ LinearOut() *synth.Wire }
	type logOuter interface{ LogOut() *synth.Wire }
	type levelOuter interface{ LevelOut() *synth.Wire }
	switch key {
	case "linear-out":
		if x, ok := c.(linearOuter); ok {
			return x.LinearOut()
		}
	case "log-out":
		if x, ok := c.(logOuter); ok {
			return x.LogOut()
		}
	case "out-level":
		if x, ok := c.(levelOuter); ok {
			return x.LevelOut()
		}
	}
	return nil
}

func wireOut(patchName, srcName, dest string, src *synth.Wire, insts map[string]component.Component, v *synth.Voice) error {
	if src == nil {
		return nil
	}
	if dest == "main" {
		v.AddVoiceMixInput(src)
		return nil
	}
	target, sel := dest, "main"
	if i := strings.IndexByte(dest, '.'); i >= 0 {
		target, sel = dest[:i], dest[i+1:]
	}
	dst, ok := insts[target]
	if !ok {
		return fmt.Errorf("patch %s: component %q out: target %q not found", patchName, srcName, target)
	}
	in, ok := dst.(component.Inputter)
	if !ok {
		return fmt.Errorf("patch %s: component %q out: target %q does not accept inputs", patchName, srcName, target)
	}
	in.AddInput(sel, src)
	return nil
}

File: c/src/patch/patch.c
#include "ondes/patch.h"

#include <ctype.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#include "ondes/component.h"
#include "ondes/spec.h"

// Defined in yaml_spec.c.
Spec *yaml_parse(Arena *a, char *mutable_buf, char *err, size_t errlen);

struct Patch {
    char *name;
    Arena arena; // owns the spec tree and the name
    Spec *root;  // top-level map: name -> component spec
    char **order; // sorted component names for deterministic wiring
    size_t n_order;
};

const char *patch_name(const Patch *p) { return p->name; }

// --- path resolution (mirrors the Go loader) ---

static void collect_dir(const char *root, char ***out, size_t *n, size_t *cap) {
    DIR *d = opendir(root);
    if (!d) return;
    struct dirent *e;
    while ((e = readdir(d))) {
        if (e->d_name[0] == '.') continue;
        char path[1024];
        snprintf(path, sizeof(path), "%s/%s", root, e->d_name);
        struct stat st;
        if (stat(path, &st) != 0) continue;
        if (S_ISDIR(st.st_mode)) {
            collect_dir(path, out, n, cap);
        } else {
            size_t len = strlen(path);
            if (len > 5 && strcmp(path + len - 5, ".yaml") == 0) {
                if (*n == *cap) {
                    *cap = *cap ? *cap * 2 : 64;
                    *out = realloc(*out, *cap * sizeof(char *));
                }
                (*out)[(*n)++] = strdup(path);
            }
        }
    }
    closedir(d);
}

static int cmp_str(const void *a, const void *b) {
    return strcmp(*(const char *const *)a, *(const char *const *)b);
}

static char *resolve_path(const char *name) {
    char **paths = NULL;
    size_t n = 0, cap = 0;
    collect_dir("program", &paths, &n, &cap);
    collect_dir("src/main/resources/program", &paths, &n, &cap);
    qsort(paths, n, sizeof(char *), cmp_str);

    char *result = NULL;
    // Exact basename match wins.
    char target[512];
    snprintf(target, sizeof(target), "%s.yaml", name);
    for (size_t i = 0; i < n && !result; i++) {
        const char *base = strrchr(paths[i], '/');
        base = base ? base + 1 : paths[i];
        if (strcmp(base, target) == 0) result = strdup(paths[i]);
    }
    // Otherwise first case-insensitive substring match.
    if (!result) {
        char lname[512];
        size_t j = 0;
        for (; name[j] && j < sizeof(lname) - 1; j++)
            lname[j] = (char)tolower((unsigned char)name[j]);
        lname[j] = 0;
        for (size_t i = 0; i < n && !result; i++) {
            char low[1024];
            size_t k = 0;
            for (; paths[i][k] && k < sizeof(low) - 1; k++)
                low[k] = (char)tolower((unsigned char)paths[i][k]);
            low[k] = 0;
            size_t plen = strlen(paths[i]);
            if (strstr(low, lname) && plen > 5 &&
                strcmp(paths[i] + plen - 5, ".yaml") == 0)
                result = strdup(paths[i]);
        }
    }
    for (size_t i = 0; i < n; i++) free(paths[i]);
    free(paths);
    return result;
}

// --- load ---

Patch *patch_load(const char *name, char *err, size_t errlen) {
    char *path = resolve_path(name);
    if (!path) {
        if (err) snprintf(err, errlen, "patch %s not found", name);
        return NULL;
    }
    FILE *f = fopen(path, "rb");
    if (!f) {
        if (err) snprintf(err, errlen, "cannot open %s", path);
        free(path);
        return NULL;
    }
    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    char *buf = malloc((size_t)sz + 1);
    size_t rd = fread(buf, 1, (size_t)sz, f);
    buf[rd] = 0;
    fclose(f);
    free(path);

    Patch *p = calloc(1, sizeof(*p));
    arena_init(&p->arena);
    p->name = arena_strdup(&p->arena, name);
    p->root = yaml_parse(&p->arena, buf, err, errlen);
    free(buf);
    if (!p->root) {
        arena_free(&p->arena);
        free(p);
        return NULL;
    }

    // Build the sorted component-name order.
    p->n_order = p->root->n_entries;
    p->order = arena_alloc(&p->arena, p->n_order * sizeof(char *));
    for (size_t i = 0; i < p->n_order; i++)
        p->order[i] = (char *)p->root->entries[i].key;
    qsort(p->order, p->n_order, sizeof(char *), cmp_str);
    return p;
}

void patch_free(Patch *p) {
    if (!p) return;
    arena_free(&p->arena);
    free(p);
}

// --- apply ---

static const Spec *spec_for(const Patch *p, const char *name) {
    return spec_get(p->root, name);
}

// wire_out connects src to the destination named by `dest`.
static int wire_out(const char *patch_name, const char *src_name,
                    const char *dest, Wire *src, Voice *v) {
    if (!src) return 0;
    if (strcmp(dest, "main") == 0) {
        voice_add_voice_mix_input(v, src);
        return 0;
    }
    char target[256];
    const char *sel = "main";
    const char *dot = strchr(dest, '.');
    if (dot) {
        size_t tl = (size_t)(dot - dest);
        if (tl >= sizeof(target)) tl = sizeof(target) - 1;
        memcpy(target, dest, tl);
        target[tl] = 0;
        sel = dot + 1;
    } else {
        snprintf(target, sizeof(target), "%s", dest);
    }
    Component *dst = voice_component(v, target);
    if (!dst) {
        fprintf(stderr, "patch %s: component %s out: target %s not found\n",
                patch_name, src_name, target);
        return -1;
    }
    if (!component_add_input(dst, sel, src)) {
        fprintf(stderr,
                "patch %s: component %s out: target %s does not accept inputs\n",
                patch_name, src_name, target);
        return -1;
    }
    return 0;
}

// nested_out returns the `out:` field of a nested block (linear-out etc).
static const char *nested_out(const Spec *spec, const char *key) {
    const Spec *m = spec_get(spec, key);
    if (!spec_is_map(m)) return NULL;
    return spec_get_str(m, "out");
}

static int patch_apply(void *ctx, Voice *v) {
    Patch *p = ctx;
    Arena *a = voice_arena(v);

    // Pass 1: make + configure every component.
    for (size_t i = 0; i < p->n_order; i++) {
        const char *name = p->order[i];
        const Spec *spec = spec_for(p, name);
        const char *type = spec_get_str(spec, "type");
        if (!type) {
            fprintf(stderr, "patch %s: component %s missing type\n", p->name,
                    name);
            return -1;
        }
        Component *c = component_make(a, type);
        if (!c) {
            fprintf(stderr, "patch %s: component %s: unknown type %s\n", p->name,
                    name, type);
            return -1;
        }
        if (component_configure(c, spec, v, name) != 0) {
            fprintf(stderr, "patch %s: component %s: configure failed\n",
                    p->name, name);
            return -1;
        }
        voice_add_component(v, name, c);
    }

    // Pass 2: resolve out: / linear-out: / log-out: / out-level:.
    for (size_t i = 0; i < p->n_order; i++) {
        const char *name = p->order[i];
        const Spec *spec = spec_for(p, name);
        Component *c = voice_component(v, name);

        const char *dest = spec_get_str(spec, "out");
        if (dest && wire_out(p->name, name, dest, component_output(c), v) != 0)
            return -1;

        const char *keys[] = {"linear-out", "log-out"};
        for (int k = 0; k < 2; k++) {
            const char *nd = nested_out(spec, keys[k]);
            if (!nd) continue;
            Wire *w = component_named_output(c, keys[k]);
            if (w && wire_out(p->name, name, nd, w, v) != 0) return -1;
        }
        const char *ol = spec_get_str(spec, "out-level");
        if (ol) {
            Wire *w = component_named_output(c, "out-level");
            if (w && wire_out(p->name, name, ol, w, v) != 0) return -1;
        }
    }
    return 0;
}

SynthPatch patch_as_synth_patch(Patch *p) {
    SynthPatch sp = {.ctx = p, .apply = patch_apply};
    return sp;
}
